File: arrays_and_strings/169-majority_element.cpp

#include <iostream>
#include <vector>

using namespace std;

/*
  Given an array `nums` of size `n`, return the majority element.

  The majority element is the element that appears more than `[n / 2]` times.
  You may assume that the majority element always exists in the array.

  Example 1:
  Input: nums = [3, 2, 3]
  Output: 3
*/

// This solution is to ensure that the vector got atleast one candidate
class Solution {
public:
  int majorityElement(vector<int> &nums) {
    int candidate = -1;
    int count = 0;

    for (int num : nums) {
      if (count == 0) {
        candidate = num;
        count = 1;
      } else if (num == candidate) {
        count++;
      } else {
        count--;
      }
    }
    return candidate;
  }
};

int main() {
  vector<int> nums = {1, 1, 2, 3, 2, 2, 2, 2, 3};
  cout << Solution().majorityElement(nums);
}

File: arrays_and_strings/169-majority_element.py
from typing import List


# This solution is for the case when nums have atleast one majority element
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        candidate = -1
        count = 0
        for num in nums:
            if count == 0:
                candidate = num
                count = 1
            elif num == candidate:
                count += 1
            else:
                count -= 1
        return candidate


nums: List[int] = [1, 3, 3, 3, 2]
print(Solution().majorityElement(nums))
